en/lessons/basics/containers.md

```diff
@@ -0,0 +1,799 @@
+comma `(a, b)`.
+
+```haskell
+ghci> myTuple = (True, "hello")
+ghci> :t myTuple
+myTuple :: (Bool, [Char])
+```
+
+We can also leave a tuple field blank, and this turns it into a function. This
+technique is called tuple sectioning. This requires the langauge extension
+`TupleSections`.
+
+```haskell
+ghci> :set -XTupleSections
+ghci> :t (True,)
+(True,) :: t -> (Bool, t)
+ghci> :t (,"hello")
+(,"hello") :: t -> (t, String)
+ghci> :t (,)
+(,) :: a -> b -> (a, b)
+ghci> (,) True "hello"
+(True,"hello")
+```
+__Note__: The syntax for enabling a language extension from within ghci has the
+following form: `:set` is the keyword that lets ghci know we are setting an
+option, the `-X` prefix tells ghci we are setting a language extension, and
+this is followed by the language extension name in pascal case.
+
+### Consuming Tuples
+
+We can consume tuples by breaking them apart. The most common approach is to
+pattern match on the structure of the tuple, and then access its contents. This
+technique is nice because it generalizes to tuples of any size.
+
+```haskell
+ghci> (\(a, b) -> not a) myTuple
+False
+ghci> (\(a, b) -> b <> " world") myTuple
+"hello world"
+
+ghci> (\(a, b, c) -> a <> " " <> b <> " " <> c) ("my", "name", "is")
+"my name is"
+```
+
+With a 2-tuple there are the functions `fst` and `snd`, and they are available
+in the standard library. 2-tuples are the only arity of tuple that comes with
+this kind of built in support, so if you need anything larger be prepared to
+write your own accessor functions.
+
+```haskell
+ghci> :t fst
+fst :: (a, b) -> a
+ghci> fst myTuple
+True
+ghci> :t snd
+snd :: (a, b) -> b
+ghci> snd myTuple
+"hello"
+```
+
+### Limitations of Tuples
+
+The main limitation with tuples in Haskell is that each different length of
+tuple is a distinct data type. That means there is no general function to add an
+additional element to any tuple. Rather these kinds of functions need to be
+defined ad hoc.
+
+Here is an example of increasing the tuple's length.
+
+```haskell
+ghci> twoTupleToThreeTuple c (a, b) = (a, b, c)
+ghci> twoTupleToThreeTuple () myTuple
+(1, "world", ())
+```
+
+Trying to call our function on a tuple of the incorrect length results in a type
+error telling us just as much; the function expected a 2-tuple but we gave it a
+3-tuple.
+
+```haskell
+ghci> twoTupleToThreeTuple True (1, 2, 3)
+
+<interactive>:19:27: error:
+    • Couldn't match expected type: (a, b)
+                  with actual type: (a0, b0, c0)
+                  ...
+```
+
+## Lists
+
+From a usability perspective, Lists solve the problem of extensibility that
+tuples face, but they can only contain one type (in other words they are
+homogeneous). Lists also have a special built in syntax.
+
+```haskell
+ghci> [1,2,3,4]
+[1,2,3,4]
+ghci> ["hello", True]
+
+<interactive>:22:11: error:
+    • Couldn't match type ‘Bool’ with ‘[Char]’
+      Expected: String
+        Actual: Bool
+    ...
+```
+
```

Comment:
I’m still confused  as to why would you put it here. There will be a chapter on user-defined datatypes, but I don’t think you need this intro here. And I don’t think it helps to understand what follows in any way.

Reply:
In another comment someone asked to see the error of having multiple types in a list. I think we are getting to a point where different opinions are at play. Would be interested what other people think about this, and I am happy to remove if that's the consensus.

Reply:
I feel that, for beginners, we should try to give a terse overview of the language that gives them the impression that they could use the language themselves. Showing too many type errors may be off-putting as it slows down the pace of progress they feel they make and their sense of understanding of the language.

Reply:
I agree with @TrueBoxGuy on this one, the pedagogical mechanics of "Do this in order to have that" (what we do) and "Here is how it can go wrong" are quite different (and the latter would be more suited for a bigger format, like a dedicated blog post)

Reply:
Friends, I'm sorry: my comment was meant for the next section where a manual definition of list is given. I don't mind the examples of errors.

Reply:
Ahhh, I was trying to demonstrate how the list data type works under the hood (since the built in syntax kind of hides that). I thought lists were a good moment to introduce inductive / recursive types. If you think that is valuable, but wasn't clear, I can try and make it more explicit. If you think that this topic is out of place and distracting, I am happy to move it.

Reply:
Yeah, i was arguing for the latter, that is: not having it in this document at all because user-defined types is a huge topic on itself. Any online tutorial on basic data structures I can remember  (tuples/lists first and foremost, and not only in Haskell, other languages too) shoes how to use lists but not how to define them. Understanding of data-definitions shouldn't be required to understand how to use lists in Haskell! Data-definitions should be the next step instead.

Reply:
/cc @Kleidukos in case you missed this discussion about list definition.

Reply:
We better take this discussion on a more appropriate medium like the Slack